File: src/test/java/com/heekng/celloct/repository/MemberRepositoryTest.java
package com.heekng.celloct.repository;

import com.heekng.celloct.entity.Member;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.transaction.annotation.Transactional;


import javax.persistence.EntityManager;

import java.util.List;
import java.util.Optional;

import static org.assertj.core.api.Assertions.*;
import static org.junit.jupiter.api.Assertions.*;

@SpringBootTest
@Transactional
class MemberRepositoryTest {

    @Autowired
    MemberRepository memberRepository;
    @Autowired
    EntityManager em;

    @Test
    @DisplayName("save 테스트")
    void save() throws Exception {
        Member member = Member.builder()
                .name("member")
                .email("member@test.com")
                .build();

        memberRepository.save(member);
        em.flush();
        em.clear();

        Member findMember = memberRepository.findById(member.getId()).get();

        assertThat(member.getId()).isEqualTo(findMember.getId());
        assertThat(member.getName()).isEqualTo(findMember.getName());
    }

    @Test
    @DisplayName("delete 테스트")
    void delete() throws Exception {
        Member member = Member.builder()
                .name("member")
                .email("member@gmail.com")
                .build();

        memberRepository.save(member);
        em.flush();
        em.clear();

        Member findMember = memberRepository.findById(member.getId()).get();
        memberRepository.delete(findMember);
        em.flush();
        em.clear();

        Optional<Member> memberOptional = memberRepository.findById(member.getId());

        assertThat(memberOptional).isEmpty();
    }

    @Test
    @DisplayName("read 테스트")
    void read() throws Exception {
        Member member1 = Member.builder()
                .name("member1")
                .email("member1@gmail.com")
                .build();

        Member member2 = Member.builder()
                .name("member2")
                .email("member2@gmail.com")
                .build();

        memberRepository.save(member1);
        memberRepository.save(member2);
        em.flush();
        em.clear();

        List<Member> members = memberRepository.findAll();

        assertThat(members.get(0).getId()).isEqualTo(member1.getId());
        assertThat(members.get(1).getId()).isEqualTo(member2.getId());
        assertThat(members.size()).isEqualTo(2);
    }

    @Test
    void findByEmailTest() throws Exception {
        //given
        Member member1 = Member.builder()
                .name("member1")
                .email("member1@gmail.com")
                .build();

        memberRepository.save(member1);
        em.flush();
        em.clear();
        //when
        Member findMember = memberRepository.findByEmail(member1.getEmail()).get();
        //then
        assertThat(findMember.getId()).isEqualTo(member1.getId());

    }

}

File: src/test/kotlin/com/heekng/celloct/repository/MemberRepositoryTest.kt
package com.heekng.celloct.repository

import com.heekng.celloct.entity.Member
import com.heekng.celloct.util.findByIdOrThrow
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.data.repository.findByIdOrNull
import org.springframework.transaction.annotation.Transactional
import javax.persistence.EntityManager

@SpringBootTest
@Transactional
class MemberRepositoryTest @Autowired constructor(
    val memberRepository: MemberRepository,
    val em: EntityManager
) {

    @Test
    fun saveTest() {
        //given
        val member = Member.fixture("member", "member@test.com")
        memberRepository.save(member)
        em.flush()
        em.clear()
        //when
        val findMember = memberRepository.findByIdOrThrow(member.id)
        //then
        assertThat(findMember.id).isEqualTo(member.id)
    }

    @Test
    fun deleteTest() {
        //given
        val member = Member.fixture("member", "member@gmail.com")
        memberRepository.save(member)
        em.flush()
        em.clear()
        //when
        val findMember = memberRepository.findByIdOrThrow(member.id)
        memberRepository.delete(findMember)
        em.flush()
        em.clear()
        //then
        val memberOrNull = memberRepository.findByIdOrNull(member.id)
        assertThat(memberOrNull).isNull()
    }

    @Test
    fun readTest() {
        //given
        val member1 = Member.fixture("member1", "member1@gmail.com")
        val member2 = Member.fixture("member2", "member2@gmail.com")
        memberRepository.save(member1)
        memberRepository.save(member2)
        em.flush()
        em.clear()
        //when
        val members = memberRepository.findAll()
        //then
        assertThat(members).hasSize(2)
        assertThat(members).extracting("name").containsExactlyInAnyOrder("member1", "member2")
    }

    @Test
    fun findByEmailTest() {
        //given
        val member = Member.fixture("member1", "member1@gmail.com")
        memberRepository.save(member)
        em.flush()
        em.clear()
        //when
        val findMember = memberRepository.findByEmail(member.email)
        //then
        assertThat(findMember!!.id).isEqualTo(member.id)
    }
}